File: src/main/java/com/kritsn/leetCodeJava/medium/_238ProductOfArrayExceptSelf.java
package com.kritsn.leetCodeJava.medium;

import java.util.Arrays;

/**
 * Copyright © 2026 Kritsn LLP. All rights reserved.
 *
 * @author Radhey (hr-sh)
 * @since Jun 14, 2026
 */
/**
 * Given an integer array nums, return an array answer such that answer[i] is equal to the
 * product of all the elements of nums except nums[i].
 *
 * The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 * You must write an algorithm that runs in O(n) time and without using the division operation.
 */
public class _238ProductOfArrayExceptSelf {

    ///////////////////////////////////////////////////////////////////////////
    // Two-pass approach:
    // 1. First pass (left to right): Fill result[i] with product of all elements to the left.
    // 2. Second pass (right to left): Multiply result[i] with product of all elements to the right.
    //
    // 🪜 Steps:
    // - Initialize answer[] with all 1s.
    // - Compute left products in first pass.
    // - Compute right products in second pass while updating answer[].
    ///////////////////////////////////////////////////////////////////////////
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] answer = new int[n];

        int leftProduct = 1;
        for (int i = 0; i < n; i++) {
            answer[i] = leftProduct;
            leftProduct *= nums[i];
        }

        int rightProduct = 1;
        for (int i = n - 1; i >= 0; i--) {
            answer[i] *= rightProduct;
            rightProduct *= nums[i];
        }

        return answer;
    }

    public int[] productExceptSelfWrong(int[] nums) {
        Long product = 1L;
        int[] res = new int[nums.length];
        Boolean[] zeros = new Boolean[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                product *= nums[i];
                zeros[i] = false;
            } else  {
                zeros[i] = true;
            }
        }
        for (int i = 0; i < nums.length; i++) {
            if (zeros[i]) {
                res[i] = Integer.parseInt((product ) + "");
            } else {
                res[i] = Integer.parseInt((product / nums[i]) + "");
            }
        }
        return res;
    }

    public static void main(String[] args) {
        _238ProductOfArrayExceptSelf solver = new _238ProductOfArrayExceptSelf();

        // Test Case 1: General case
        int[] nums1 = {1, 2, 3, 4};
        System.out.println("Test Case 1: nums = " + Arrays.toString(nums1)
                + " -> answer = " + Arrays.toString(solver.productExceptSelf(nums1)));
        // Expected: [24, 12, 8, 6]

        // Test Case 2: Includes a zero
        int[] nums2 = {1, 2, 0, 4};
        System.out.println("Test Case 2: nums = " + Arrays.toString(nums2)
                + " -> answer = " + Arrays.toString(solver.productExceptSelf(nums2)));
        // Expected: [0, 0, 8, 0]

        // Test Case 3: All same elements
        int[] nums3 = {2, 2, 2, 2};
        System.out.println("Test Case 3: nums = " + Arrays.toString(nums3)
                + " -> answer = " + Arrays.toString(solver.productExceptSelf(nums3)));
        // Expected: [8, 8, 8, 8]

        // Test Case 4: Two elements
        int[] nums4 = {10, 1};
        System.out.println("Test Case 4: nums = " + Arrays.toString(nums4)
                + " -> answer = " + Arrays.toString(solver.productExceptSelf(nums4)));
        // Expected: [1, 10]

        // Test Case 5: Contains multiple zeros
        int[] nums5 = {0, 0, 3, 4};
        System.out.println("Test Case 5: nums = " + Arrays.toString(nums5)
                + " -> answer = " + Arrays.toString(solver.productExceptSelf(nums5)));
        // Expected: [0, 0, 0, 0]
    }
}

File: src/main/java/com/kritsn/leetcode/_238ProductOfArrayExceptSelf.kt
package com.kritsn.leetcode

///////////////////////////////////////////////////////////////////////////
//           Copyright © 2025 Kritsn LLP. All rights reserved.
//                      @author Radhey (hr-sh)
//                      @since Jul 12, 2025
///////////////////////////////////////////////////////////////////////////

/**
 * Given an integer array nums, return an array answer such that answer[i] is equal to the
 * product of all the elements of nums except nums[i].
 *
 * The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 * You must write an algorithm that runs in O(n) time and without using the division operation.
 */
class _238ProductOfArrayExceptSelf {

    ///////////////////////////////////////////////////////////////////////////
    // Two-pass approach:
    // 1. First pass (left to right): Fill result[i] with product of all elements to the left.
    // 2. Second pass (right to left): Multiply result[i] with product of all elements to the right.
    //
    // 🪜 Steps:
    // - Initialize answer[] with all 1s.
    // - Compute left products in first pass.
    // - Compute right products in second pass while updating answer[].
    ///////////////////////////////////////////////////////////////////////////
    fun productExceptSelf(nums: IntArray): IntArray {
        val n = nums.size
        val answer = IntArray(n) { 1 }

        var leftProduct = 1
        for (i in 0 until n) {
            // Set current answer as product of all elements to the left of i
            answer[i] = leftProduct
            leftProduct *= nums[i] // Update left product
        }

        var rightProduct = 1
        for (i in n - 1 downTo 0) {
            // Multiply with product of all elements to the right of i
            answer[i] *= rightProduct
            rightProduct *= nums[i] // Update right product
        }

        return answer
    }
}

// 🔍 Main method with clearly labeled test cases
fun main() {
    val solver = _238ProductOfArrayExceptSelf()

    // Test Case 1: General case
    val nums1 = intArrayOf(1, 2, 3, 4)
    println("Test Case 1: nums = ${nums1.contentToString()} -> answer = ${solver.productExceptSelf(nums1).contentToString()}") // Expected: [24,12,8,6]

    // Test Case 2: Includes a zero
    val nums2 = intArrayOf(1, 2, 0, 4)
    println("Test Case 2: nums = ${nums2.contentToString()} -> answer = ${solver.productExceptSelf(nums2).contentToString()}") // Expected: [0,0,8,0]

    // Test Case 3: All same elements
    val nums3 = intArrayOf(2, 2, 2, 2)
    println("Test Case 3: nums = ${nums3.contentToString()} -> answer = ${solver.productExceptSelf(nums3).contentToString()}") // Expected: [8,8,8,8]

    // Test Case 4: Single element (edge case, but valid assumption that len >= 2 in Leetcode)
    val nums4 = intArrayOf(10, 1)
    println("Test Case 4: nums = ${nums4.contentToString()} -> answer = ${solver.productExceptSelf(nums4).contentToString()}") // Expected: [1,10]

    // Test Case 5: Contains multiple zeros
    val nums5 = intArrayOf(0, 0, 3, 4)
    println("Test Case 5: nums = ${nums5.contentToString()} -> answer = ${solver.productExceptSelf(nums5).contentToString()}") // Expected: [0,0,0,0]
}
